### PR TITLE
fix: reject monitors with invalid relabel configs

### PR DESCRIPTION
**Pull request by simonpasquier**
## Description

I noticed that probes, service monitors, pod monitors and scrape configs with invalid relabel configs were not rejected by the operator. This PR fixes it.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [X] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
reject ServiceMonitor, PodMonitor, Probe and ScrapeConfig with invalid relabel configs.
```

### CODE REVIEW

File: pkg/prometheus/resource_selector.go [hunk 1]
```diff
@@ -225,55 +219,75 @@ func validateScrapeIntervalAndTimeout(p monitoringv1.PrometheusInterface, scrape
...
+
 	version, err := semver.ParseTolerant(promVersion)
 	if err != nil {
 		return errors.Wrap(err, "failed to parse Prometheus version")
 	}
+
 	minimumVersionCaseActions := version.GTE(semver.MustParse("2.36.0"))
 	minimumVersionEqualActions := version.GTE(semver.MustParse("2.41.0"))
+	if rc.Action == "" {
+		rc.Action = string(relabel.Replace)
```
Comment (slashpai) [inline pkg/prometheus/resource_selector.go:248]: shouldn't this be already taken care since we set default as `replace`?
https://github.com/prometheus-operator/prometheus-operator/blob/78a28753bef7f702c27f16ea38449f599d805301/pkg/apis/monitoring/v1/prometheus_types.go#L1340-L1341

File: pkg/prometheus/resource_selector_test.go [hunk 2]
```diff
@@ -400,77 +415,193 @@ func TestValidateRelabelConfig(t *testing.T) {
...
+				}
+			},
+			selected: false,
 		},
 	} {
-		t.Run(fmt.Sprintf("case %s %s", tc.scenario, tc.proberSpec.URL), func(t *testing.T) {
-			err := validateProberURL(tc.proberSpec.URL)
-			if err != nil && !tc.expectedErr {
-				t.Fatalf("expected no error, got: %v", err)
+		t.Run(fmt.Sprintf(tc.scenario), func(t *testing.T) {
```
Comment (ArthurSens) [inline pkg/prometheus/resource_selector_test.go:570]:
```suggestion
		t.Run(tc.scenario, func(t *testing.T) {
```

**Review by ArthurSens — COMMENTED** [review]
I'm struggling a little bit to understand what/where is the bug that you're fixing here 😬 

Do I understand it correctly that we're allowing actions with a mixture of Lower/Upper case and if we fail to compare with `relabel.<action name>`?

If that's the case, what do you think about passing this validation to kubebuilder annotations?

File: pkg/prometheus/resource_selector_test.go (hunk 2 above)
Comment (ArthurSens) [reply pkg/prometheus/resource_selector_test.go:570]: small nit on all tests created

File: pkg/prometheus/resource_selector.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/prometheus/resource_selector.go:248]: This is for tests + environments that don't have a recent version of the CRD (e.g. one without the default value).

**Comment by simonpasquier** [comment]
> I'm struggling a little bit to understand what/where is the bug that you're fixing here 😬
> 


One issue is here:

https://github.com/prometheus-operator/prometheus-operator/blob/34bb905e3ca90963ffdaa50cbe1924137f9a5ed7/pkg/prometheus/resource_selector.go#L535-L548

If an invalid relabel config is detected, it will be logged but it won't continue from the outer for loop so the probe will be added to the map.

Another issue here:

https://github.com/prometheus-operator/prometheus-operator/blob/34bb905e3ca90963ffdaa50cbe1924137f9a5ed7/pkg/prometheus/resource_selector.go#L155-L172

If L157 returns an error, `err` might be overwritten at L165 and again the service monitor isn't rejected...

TL;DR: write unit tests to assert that the code is doing what it should ;)

> Do I understand it correctly that we're allowing actions with a mixture of Lower/Upper case and then we fail to compare with `relabel.<action name>`?
> 
> If that's the case, what do you think about passing this validation to kubebuilder annotations?

No upper/lower case actions are fine (in fact the Prometheus config supports everything but will always compare to the lower-case version).

File: pkg/prometheus/resource_selector.go [hunk 3]
```diff
@@ -152,20 +152,14 @@ func (rs *ResourceSelector) SelectServiceMonitors(ctx context.Context, listFn Li
 				break
 			}
 
-			for _, rl := range endpoint.RelabelConfigs {
-				if rl.Action != "" {
```
Comment (simonpasquier) [inline pkg/prometheus/resource_selector.go:156]: I'm not sure why we had this test in the first place. I've modified `validateRelabelConfig()` to default to the replace action if the field is empty.

**Review by ArthurSens — APPROVED** [review]

**Comment by ArthurSens** [comment]
> > I'm struggling a little bit to understand what/where is the bug that you're fixing here 😬
> 
> One issue is here:
> 
> https://github.com/prometheus-operator/prometheus-operator/blob/34bb905e3ca90963ffdaa50cbe1924137f9a5ed7/pkg/prometheus/resource_selector.go#L535-L548
> 
> If an invalid relabel config is detected, it will be logged but it won't continue from the outer for loop so the probe will be added to the map.
> 
> Another issue here:
> 
> https://github.com/prometheus-operator/prometheus-operator/blob/34bb905e3ca90963ffdaa50cbe1924137f9a5ed7/pkg/prometheus/resource_selector.go#L155-L172
> 
> If L157 returns an error, `err` might be overwritten at L165 and again the service monitor isn't rejected...
> 
> TL;DR: write unit tests to assert that the code is doing what it should ;)
> 
> > Do I understand it correctly that we're allowing actions with a mixture of Lower/Upper case and then we fail to compare with `relabel.<action name>`?
> > If that's the case, what do you think about passing this validation to kubebuilder annotations?
> 
> No upper/lower case actions are fine (in fact the Prometheus config supports everything but will always compare to the lower-case version).

Ooooh now I get it, we're breaking the inner loop when we wanted to break the outer loop. Thanks for the explanation!
